### PR TITLE
Required support

### LINKED ISSUE #84: Add required property to element
We use the element for agreements on our checkout process, and a checkbox is always required to be filled in. We'd like to be able to make the field required and add an error message when the customer doesn't fill it in.

### PR DESCRIPTION
**Pull request by NickdeK**
Fixes #84 

Add required property and validation functions.

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-checkbox/85)

### CODE REVIEW

**Comment by jouni** [comment]
Thanks for the PR!

I’m wondering about the use cases – do you have some other use case than the common “I agree to the terms” type of checkbox in most registration forms?

Regarding the code, we don’t currently have any field borders defined in Lumo, so I would like to avoid them here as well. The error-10pct background color should suffice like it’s in vaadin-text-field.

**Comment by web-padawan** [comment]
Per [whatwg spec](https://html.spec.whatwg.org/multipage/input.html#concept-input-apply), we should add the support for this feature. 

Use cases might vary, e. g. imagine some "installation wizard" where the installable parts of the software are marked by checkboxes and at least one of those is likely to be required (although it is usually disabled at the same time).

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85#-LACMli_6unIMTX8MZew:-LACMli_6unIMTX8MZex:b-mndmya) (no related file):*
Since this is a new feature, it would be nice to have a demo for it as well. It might be that we also want to demonstrate usage with `

---

*[src/vaadin-checkbox.html, line 262 at r1](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85#-LACHi9v5rcO8HnqjzCt:-LACHi9v5rcO8HnqjzCu:b-i8nun6) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/2cefc4ed869567489ecb2b09c7c53ae9874148f9/src/vaadin-checkbox.html#L262)):*
> ```HTML
>         }
> 
>         get validateElement() {
> ```

What's the point of the inroducing this API?
We already have the reference to the native checkbox stored in the `_nativeCheckbox` property

---

*[src/vaadin-checkbox.html, line 298 at r1](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85#-LACHt5m7BhjdjEWQKTO:-LACHt5m7BhjdjEWQKTP:b-owqws9) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/2cefc4ed869567489ecb2b09c7c53ae9874148f9/src/vaadin-checkbox.html#L298)):*
> ```HTML
>         checkValidity() {
>           if (this.required) {
>             return this.validateElement.checkValidity();
> ```

We need to double check whether this works nice in Safari 9

---

*[src/vaadin-checkbox.html, line 300 at r1](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85#-LACJXRb0Y0821lj4vQK:-LACJXRb0Y0821lj4vQL:bah48k5) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/2cefc4ed869567489ecb2b09c7c53ae9874148f9/src/vaadin-checkbox.html#L300)):*
> ```HTML
>             return this.validateElement.checkValidity();
>           } else {
>             return !this.invalid;
> ```

This code introduces the following issue:

1. Add the checkbox, which is not required but invalid

```
<vaadin-checkbox checked invalid>Option label</vaadin-checkbox>
```
2. Uncheck the checkbox - it stays invalid

I recall us having similar kind of the code in text-field, which was a problem for one of the components using it internally. 

So, for checkbox, we should probably consider making `invalid` a readonly property, and only set it to true if the checkbox is required and unchecked.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85#-:-LACMN4aE9lECkolbYnI:bsftmbe)*

**Comment by web-padawan** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85#-LACMli_6unIMTX8MZew:-LACNoCE4vkmXtFcZR88:b-9kqe09) (no related file):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

Since this is a new feature, it would be nice to have a demo for it as well. It might be that we also want to demonstrate usage with `
</blockquote></details>

I meant that we should demonstrate the usage with iron-form (reviewable doesn't like backticks).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85)*

**Comment by jouni** [comment]
> Use cases might vary, e. g. imagine some "installation wizard" where the installable parts of the software are marked by checkboxes and at least one of those is likely to be required (although it is usually disabled at the same time).

Yes, I wasn’t trying to dismiss the feature completely. Just started wondering what are the use cases for a *single* checkbox to be required, and if it’s more common for a group of checkboxes to be required and if this feature should belong to a vaadin-checkbox-group element instead. But that doesn’t really seem viable, as you couldn’t specify exactly which or how many of those checkboxes should be checked to pass the validation.

**Comment by NickdeK** [comment]
I just copied over all the related functions from the vaadin-text-field-mixin, of course I didn't check every scenario just the one we needed right now ;)

Also I forgot to add a error message to the solution, so that needs to be added I guess.

**Comment by web-padawan** [comment]
>  Just started wondering what are the use cases for a single checkbox to be required, and if it’s more common for a group of checkboxes to be required and if this feature should belong to a vaadin-checkbox-group element instead

The spec only states that select element and radio-button group behave differently: https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-being-missing

There is no "checkbox group" entity in the web currently and I'm not sure if we want to introduce it as a separate component, but the standalone checkbox can be required by itself.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/vaadin-checkbox/85#-:-LACSXu52ngmAggVtCU1:b-eenmd)*

**Comment by jouni** [comment]
> the standalone checkbox can be required by itself.

Yes, I do agree.

> There is no "checkbox group" entity in the web currently and I'm not sure if we want to introduce it as a separate component

Well, this has been a requirement in at least Expense Manager, and Vaadin Framework 8 and earlier have a CheckBoxGroup component, previously called OptionGroup which could be multi-select or not mapping to checkboxes and radio buttons.

But not in this PR :)

**Comment by mstahv** [comment]
Thanks @NickdeK for providing this feature! 

I'd actually need this feature for a Vaadin 10 demo, but as it is not merged yet, I'll just have to go on with hacks :-(

**Comment by NickdeK** [comment]
Right! I'm not really sure what is needed to prep this PR further so it can be merged. If there are some clear tasks to be done (code changes) I wouldn't mind doing those to increase the change for this PR to be merged :)

**Comment by jouni** [comment]
I suppose this has just fallen through the cracks by the team. Thanks for bumping it.

**Comment by tomivirkki** [comment]
Overall the code looks good to me. What's missing are the tests. All the necessary tests should be available in other components that implement this (required) feature. Duplicating them here should be enough.

File: theme/lumo/vaadin-checkbox.html [hunk 1]
```diff
@@ -117,6 +117,12 @@
       :host([indeterminate][disabled]) [part="checkbox"]::after {
         background-color: var(--lumo-contrast-30pct);
       }
+
+      /* Invalid */
+      :host([invalid]) [part="checkbox"] {
+        background-color: var(--lumo-error-color-10pct);
+        border-color: var(--lumo-error-color-50pct);
```
Comment (jouni) [inline theme/lumo/vaadin-checkbox.html]: Remove the border-color. I don’t see it being used anywhere.

**Comment by jouni** [comment]
Not sure if we need an example for this. I would say not immediately, at least, as this is not perhaps that common use case and easy to discover when needed.

**Comment by mstahv** [comment]
I think it is very common use case. Many forms have checkbox to agree terms or something similar that needs to be checked before one can submit. But I do agree it is easy to find. I was trying to find it, but it just didn't exists :-) Found it when was about to create an enhancement issues.

**Comment by limonte** [comment]
@jouni please add clear instructions on how to proceed with this PR. There are many questions and doubts from you, some of them need answers and some of them not. I guess @NickdeK is confused about how he should proceed.

Is "Remove the border-color. I don’t see it being used anywhere." the only one requirement?

**Comment by jouni** [comment]
> Is "Remove the border-color. I don’t see it being used anywhere." the only one requirement?

The border, and then an agreement when to run validation. My proposal is that we validate on blur as well as when the value is toggled.

**Comment by NickdeK** [comment]
- [x] Remove border color
- [ ] Run validation on blur or when value is toggled
- [ ] Add tests 

Are you discussing the validation method for this? I can imagine that you want this on toggle because onBlur might give the situation where the checkbox is in a invalid state and stays on the invalid state when you check it, the invalid state will disappear when you click on a other element? Not sure about this tho.

**Comment by jouni** [comment]
Yeah, I think we are on the same page:

- If the user focuses a required checkbox with the keyboard and then blurs it without checking it, we should mark it invalid
- When the user toggles the checkbox (either with the mouse or keyboard) we should run validation and remove/add the invalid flag (i.e. remove the invalid state immediately when the user toggles it)

I assume some apps might have some other UI logic based on the invalid state of a checkbox, and it would be ideal to update those parts of the UI immediately when the user clicks the checkbox and not just after they blur it (click somewhere else).

**Comment by NickdeK** [comment]
Right, so I just have added the requested on blur validation check and the tests but I have some problems with the on blur validation. Validation happens on the validatable element using the browser validation methods and it looks like the browser doesn't require a checkbox to be checked on blur when no interaction has happened on the checkbox yet, so it won't invalidate in that case and the specific test is now failing.

I'm not sure how to proceed here; do we want to overwrite normal browser validation behaviour and check if the checkbox is checked on blur? Or am I missing something here and should it validate on blur by browser standards?

**Comment by tomivirkki** [comment]
Looks like this has been stuck for quite some time, sorry about that. Is this feature still relevant for a single checkbox now that checkbox group (with required support) is available?

**Comment by NickdeK** [comment]
I'm not sure. You have to check that out yourselves I'm afraid; I've already putten in to much time in this PR and I'm not willing to devote more time for this. Good luck!

**Comment by knoobie** [comment]
> Is this feature still relevant for a single checkbox now that checkbox group (with required support) is available?

Yes, it is. A single checkbox has a different Look and Feel, usability and accessibility. 

For Example: A single checkbox is not un-checkable default (see https://vaadin.com/components/vaadin-radio-button/html-examples/radio-button-basic-demos and try to un-check english, after you select it)

**Comment by web-padawan** [comment]
Note for anyone taking care of further implementation: one of the reasons why we haven't merged this PR so far is that we are missing the solution for single checkbox required indicator.

This was especially requested in https://github.com/vaadin/vaadin-checkbox/issues/133#issue-380595598

**Comment by web-padawan** [comment]
It has been a while since this PR was submitted. Apologies about keeping it open for so long without any progress.

The component has been converted to Polymer 3 so there are merge commits that would be tricky to solve.
Also, the new development for Vaadin web components will happen in the [monorepo](https://github.com/vaadin/vaadin-web-components).

I'm closing this now. We will submit a new PR to the monorepo when the linked issue is prioritised.

**Pull request closed without merging**